proto/v1beta1/intoto.proto

```diff
@@ -0,0 +1,135 @@
+// in-toto layout.
+message SigningKey {
+  // key_id is an identifier for the signing key.
+  string key_id = 1;
+
+  // This field identifies the hash algorithms used in the particular key ID.
+  repeated string key_id_hash_algorithms = 2;
+
+  // This field identifies the specific signing method. Eg: "RSASSA-PSS",
+  // "ed25519", and "ecdsa".
+  string key_type = 3;
+
+  // This field contains the actual public key.
+  string public_key_value = 4;
+
+  // This field contains the overall scheme used for the creation of the signing
+  // key. Eg: "rsassa-pss-sha256".
+  string key_scheme = 5;
+}
+
+// This corresponds to a signed in-toto attestation - it is made up of one or
+// more signatures and the link attestation itself. This is used for occurrences
+// of a Grafeas in-toto note.
+message Details {
+  repeated Signature signatures = 1;
+  Link link = 2 [json_name="signed"];
+}
+
+// A signature object consists of the KeyID used and the signature itself.
+message Signature {
+  string key_id = 1 [json_name="keyid"];
+  string signature = 2 [json_name="sig"];
+}
+
+// This corresponds to an in-toto link attestation.
+message Link {
+  // This field contains the full command executed for the step. This can also
+  // be empty if attestations are generated for operations that aren't directly
+  // mapped to a specific command. Each term in the command is an independent
+  // string in the list. An example of a command in the in-toto metadata field
+  // is:
+  // "command": ["git", "clone", "https://github.com/in-toto/demo-project.git"]
+  repeated string effective_command = 1 [json_name="command"];
+
+  // Materials are the supply chain artifacts that go into the step and are used
+  // for the operation performed. The key of the map is the path of the artifact
+  // and the structure contains the recorded hash information. An example is:
+  // "materials": {
+  //   "foo/bar": {
+  //     "sha256": "ebebf..."
+  //   }
+  // }
+  map<string, google.protobuf.Struct> materials = 2;
+
+  // Products are the supply chain artifacts generated as a result of the step.
+  // The structure is identical to that of materials.
+  map<string, google.protobuf.Struct> products = 3;
+
+  // By Products are data generated as part of a software supply chain step, but
+  // are not the actual result of the step. The suggested fields are "stderr",
+  // "stdout", and "return-value".
+  map<string, google.protobuf.Struct> byproducts = 4;
+
+  // This is a field that can be used to capture information about the
+  // environment. It is suggested for this field to contain information that
+  // details environment variables, filesystem information, and the present
+  // working directory. The recommended structure of this field is:
+  // "environment": {
+  //   "variables": "<ENV>",
+  //   "filesystem": "<FS>",
+  //   "workdir": "<CWD>",
+  //   "<ANY OTHER RELEVANT FIELDS>": "..."
+  // }
+  map<string, google.protobuf.Struct> environment = 5;
```

Comment:
ditto here for the specialized `Environment` proto message. The fields are what you already listed: `variables`, `filesystem`, `workdir` etc. And you can always add more fields to the proto message without having to release new API. However, since this is `v1beta1` API version, some of this could be ok, assuming all dependent software is updated accordingly.

Reply:
as above, those fields are just some examples we came up with, but they may not occur in practice.

Reply:
merging into 1 thread :)